File: driver-service/src/main/java/com/dulfinne/taxi/driverservice/security/AuthoritiesConverter.java
package com.dulfinne.taxi.driverservice.security;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

import com.dulfinne.taxi.driverservice.util.TokenConstants;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;

public class AuthoritiesConverter {
  private final JwtGrantedAuthoritiesConverter converter = new JwtGrantedAuthoritiesConverter();

  public Set<GrantedAuthority> convert(Jwt jwt) {
    Set<GrantedAuthority> authorities = new HashSet<>(converter.convert(jwt));

    var realmRoles = jwt.getClaimAsStringList(TokenConstants.ROLES);
    Set<GrantedAuthority> grantedAuthorities =
        realmRoles.stream()
                .map(SimpleGrantedAuthority::new)
                .collect(Collectors.toSet());

    authorities.addAll(grantedAuthorities);
    return authorities;
  }
}

File: ride-service/src/main/kotlin/com/dulfinne/taxi/rideservice/security/AuthoritiesConverter.kt
package com.dulfinne.taxi.rideservice.security

import com.dulfinne.taxi.rideservice.util.TokenConstants
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
import org.springframework.security.core.authority.SimpleGrantedAuthority
import java.util.stream.Collectors

class AuthoritiesConverter {
    private val converter = JwtGrantedAuthoritiesConverter()

    fun convert(jwt: Jwt): Set<GrantedAuthority> {
        val authorities = HashSet(converter.convert(jwt))

        val realmRoles = jwt.getClaimAsStringList(TokenConstants.ROLES_CLAIM)
        val grantedAuthorities = realmRoles.stream()
            .map { SimpleGrantedAuthority(it) }
            .collect(Collectors.toSet())

        authorities.addAll(grantedAuthorities)
        return authorities
    }
}
